File: java/src/test/java/org/htmlsanitizer/ConformanceTest.java
package org.htmlsanitizer;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.function.Consumer;

/**
 * The 12-check binding conformance suite (docs/conformance.md).
 *
 * <p>Proves the Java binding marshals every value shape across the FFI. It is
 * NOT a sanitizer test suite — the behavioural cases live in the engine's own
 * tests and run once, in Aether.
 *
 * <h2>Why a main-method runner instead of JUnit</h2>
 * This suite must run with nothing but a JDK, so {@code java/.tests.ae} works
 * offline and without Maven resolving anything. The assertions here are a
 * dozen lines; a test framework would be the only thing in the whole binding
 * that needed a network fetch. {@code java/pom.xml} still declares JUnit 5 for
 * downstream consumers who want to run this class under Maven — the checks
 * are plain methods, so wrapping them in {@code @Test} is mechanical.
 *
 * <p>Run:
 * <pre>{@code
 * javac -d out $(find src -name '*.java')
 * HTMLSANITIZER_LIB=... java --enable-native-access=ALL-UNNAMED \
 *     -cp out org.htmlsanitizer.ConformanceTest
 * }</pre>
 */
public final class ConformanceTest {

    private static int passed = 0;
    private static final List<String> failures = new ArrayList<>();

    public static void main(String[] args) {
        // ---- the 12 required checks ----
        check("01 script removed", s ->
                assertEquals("<div>Hello  world!</div>",
                        s.sanitize("<div>Hello <script>alert(1)</script> world!</div>")));

        check("02 onclick removed", s ->
                assertEquals("<div>Hello</div>",
                        s.sanitize("<div onclick=\"alert(1)\">Hello</div>")));

        check("03 empty string", s -> assertEquals("", s.sanitize("")));

        check("04 utf-8 round trip", s ->
                assertEquals("<div>café ☕</div>", s.sanitize("<div>café ☕</div>")));

        check("05 allow custom tag", s -> {
            assertEquals("", s.sanitize("<my-widget>x</my-widget>"));
            s.allowedTags().add("my-widget");
            assertEquals("<my-widget>x</my-widget>", s.sanitize("<my-widget>x</my-widget>"));
        });

        check("06 disallow tag", s -> {
            assertEquals("<div>x</div>", s.sanitize("<div>x</div>"));
            s.allowedTags().remove("div");
            assertEquals("", s.sanitize("<div>x</div>"));
        });

        check("07 membership and count", s -> {
            assertTrue("http is allowed", s.allowedSchemes().contains("http"));
            assertTrue("gopher is not", !s.allowedSchemes().contains("gopher"));
            assertEquals(2, s.allowedSchemes().size());
        });

        check("08 enumeration", s -> {
            List<String> got = s.allowedSchemes().toList();
            Collections.sort(got);
            assertEquals(List.of("http", "https").toString(), got.toString());
        });

        check("09 keep child nodes", s -> {
            assertEquals("<div></div>",
                    s.sanitize("<div><nope>Hello <span>world</span></nope></div>"));
            s.keepChildNodes(true);
            assertTrue("flag reads back", s.keepChildNodes());
            assertEquals("<div>Hello <span>world</span></div>",
                    s.sanitize("<div><nope>Hello <span>world</span></nope></div>"));
        });

        check("10 on_removing_tag cancels", s -> {
            List<String> seen = new ArrayList<>();
            s.onRemovingTag((node, reason) -> {
                seen.add(node.name() + "/" + reason);
                return node.name().equals("keep-me");
            });
            assertEquals("<div><keep-me>a</keep-me></div>",
                    s.sanitize("<div><keep-me>a</keep-me><drop-me>b</drop-me></div>"));
            assertTrue("saw keep-me/0 in " + seen, seen.contains("keep-me/0"));
            assertTrue("saw drop-me/0 in " + seen, seen.contains("drop-me/0"));
        });

        check("11 on_filter_url rewrites", s -> {
            s.onFilterUrl((node, raw, resolved) ->
                    resolved.equals("https://example.com/logo.png")
                            ? "https://cdn.example.net/logo.png"
                            : resolved);
            assertEquals("<img src=\"https://cdn.example.net/logo.png\">",
                    s.sanitize("<img src=\"logo.png\">", "https://example.com"));
        });

        run("12 handles are independent", () -> {
            try (HtmlSanitizer a = new HtmlSanitizer(); HtmlSanitizer b = new HtmlSanitizer()) {
                a.allowedTags().add("only-in-a");
                assertTrue("a knows the tag", a.allowedTags().contains("only-in-a"));
                assertTrue("b does not", !b.allowedTags().contains("only-in-a"));
            }
        });

        // ---- extras that exercise the remaining callback shapes ----

        check("on_removing_attribute sees the attribute", s -> {
            List<String> seen = new ArrayList<>();
            s.onRemovingAttribute((elem, attr, reason) -> {
                seen.add(elem.name() + "/" + attr.name() + "/" + attr.value());
                return false;
            });
            assertEquals("<div>x</div>", s.sanitize("<div onclick=\"alert(1)\">x</div>"));
            assertTrue("saw div/onclick/alert(1) in " + seen,
                    seen.contains("div/onclick/alert(1)"));
        });

        check("on_removing_comment cancels", s -> {
            s.onRemovingComment(node -> true);
            assertEquals("<div>a<!-- keep -->b</div>", s.sanitize("<div>a<!-- keep -->b</div>"));
        });

        check("on_removing_style is four-arg", s -> {
            List<String> seen = new ArrayList<>();
            s.onRemovingStyle((elem, name, value, reason) -> {
                seen.add(name + "/" + value);
                return name.equals("-custom-thing");
            });
            String out = s.sanitize("<div style=\"-custom-thing: 3; color: red\">x</div>");
            assertTrue("kept -custom-thing, got " + out, out.contains("-custom-thing"));
            assertTrue("saw -custom-thing/3 in " + seen, seen.contains("-custom-thing/3"));
        });

        check("post_process_node visits", s -> {
            List<Integer> kinds = new ArrayList<>();
            s.onPostProcessNode(node -> kinds.add(node.kind()));
            s.sanitize("<div><span>a</span><span>b</span></div>");
            assertTrue("visited at least one node", !kinds.isEmpty());
        });

        check("node tree navigation", s -> {
            int[] captured = new int[2];
            s.onPostProcessDom(doc -> {
                captured[0] = doc.kind();
                captured[1] = doc.children().size();
            });
            s.sanitize("<div>a</div><p>b</p>");
            assertEquals(Native.NODE_DOCUMENT, captured[0]);
            assertTrue("document had >= 2 children, got " + captured[1], captured[1] >= 2);
        });

        check("abi version", s -> assertTrue("abi >= 1", s.abiVersion() >= 1));

        check("sanitize_document is wired", s ->
                assertEquals("<div>doc</div>",
                        s.sanitizeDocument("<div>doc<script>x</script></div>")));

        check("attribute set_value rewrites", s -> {
            s.onRemovingAttribute((elem, attr, reason) -> {
                if (attr.name().equals("onclick")) {
                    // Prove the setter reaches the DOM: rename the value and
                    // keep the attribute by cancelling the removal.
                    attr.setValue("safe");
                    return true;
                }
                return false;
            });
            assertEquals("<div onclick=\"safe\">x</div>",
                    s.sanitize("<div onclick=\"alert(1)\">x</div>"));
        });

        run("closed sanitizer rejects use", () -> {
            HtmlSanitizer s = new HtmlSanitizer();
            s.close();
            try {
                s.sanitize("<div>x</div>");
                throw new AssertionError("expected IllegalStateException");
            } catch (IllegalStateException expected) {
                // exactly right
            }
        });

        // ---- report ----
        System.out.println();
        if (failures.isEmpty()) {
            System.out.println("PASS — " + passed + " checks");
            System.exit(0);
        }
        System.out.println("FAIL — " + failures.size() + " of "
                + (passed + failures.size()) + " checks failed:");
        for (String f : failures) System.out.println("  " + f);
        System.exit(1);
    }

    /** Run one check against a fresh sanitizer, always closing it. */
    private static void check(String name, Consumer<HtmlSanitizer> body) {
        run(name, () -> {
            try (HtmlSanitizer s = new HtmlSanitizer()) {
                body.accept(s);
            }
        });
    }

    private static void run(String name, Runnable body) {
        try {
            body.run();
            passed++;
            System.out.println("  ok   " + name);
        } catch (Throwable t) {
            failures.add(name + ": " + t);
            System.out.println("  FAIL " + name + ": " + t);
        }
    }

    private static void assertEquals(Object expected, Object actual) {
        if (!String.valueOf(expected).equals(String.valueOf(actual))) {
            throw new AssertionError("expected <" + expected + "> but was <" + actual + ">");
        }
    }

    private static void assertTrue(String what, boolean cond) {
        if (!cond) throw new AssertionError("expected " + what);
    }

    private ConformanceTest() {
    }
}

File: clojure/test/org/htmlsanitizer/conformance_test.clj
(ns org.htmlsanitizer.conformance-test
  "The 12-check binding conformance suite (docs/conformance.md), in Clojure.

  Proves the **Clojure layer** marshals every value shape correctly. Since that
  layer sits on the Java binding rather than on its own FFI, what this suite
  really pins down is that the Clojure wrapping — `reify`d SAM callbacks,
  keyword reasons, the live allow-list fns — reaches the same engine behaviour
  the Java and Python suites see. A `reify` that returned the wrong type, or an
  `allow!` that mutated a copy instead of the live view, would fail here and
  nowhere else.

  It is NOT a sanitizer test suite — the behavioural cases live in the engine's
  own tests and run once, in Aether.

  `clojure.test` rather than an external framework, so the run needs nothing but
  the Clojure jar itself: it works offline and cannot fail resolving a
  test-framework artifact."
  (:require [clojure.test :refer [deftest is run-tests testing]]
            [org.htmlsanitizer.core :as hs])
  (:import (org.htmlsanitizer HtmlSanitizer)))

;; ---- the 12 required checks ----------------------------------------------

(deftest test-01-script-removed
  (with-open [s (hs/sanitizer)]
    (is (= "<div>Hello  world!</div>"
           (hs/sanitize s "<div>Hello <script>alert(1)</script> world!</div>")))))

(deftest test-02-onclick-removed
  (with-open [s (hs/sanitizer)]
    (is (= "<div>Hello</div>" (hs/sanitize s "<div onclick=\"alert(1)\">Hello</div>")))))

(deftest test-03-empty-string
  (with-open [s (hs/sanitizer)]
    (is (= "" (hs/sanitize s "")))))

(deftest test-04-utf8-round-trip
  (with-open [s (hs/sanitizer)]
    (is (= "<div>café ☕</div>" (hs/sanitize s "<div>café ☕</div>")))))

(deftest test-05-allow-custom-tag
  (with-open [s (hs/sanitizer)]
    (is (= "" (hs/sanitize s "<my-widget>x</my-widget>")))
    (hs/allow! s :tags "my-widget")
    (is (= "<my-widget>x</my-widget>" (hs/sanitize s "<my-widget>x</my-widget>")))))

(deftest test-06-disallow-tag
  (with-open [s (hs/sanitizer)]
    (is (= "<div>x</div>" (hs/sanitize s "<div>x</div>")))
    (hs/disallow! s :tags "div")
    (is (= "" (hs/sanitize s "<div>x</div>")))))

(deftest test-07-membership-and-count
  (with-open [s (hs/sanitizer)]
    (is (hs/allowed? s :schemes "http"))
    (is (not (hs/allowed? s :schemes "gopher")))
    (is (= 2 (hs/allow-count s :schemes)))))

(deftest test-08-enumeration
  (with-open [s (hs/sanitizer)]
    (is (= #{"http" "https"} (hs/allowed s :schemes)))))

(deftest test-09-keep-child-nodes
  (with-open [s (hs/sanitizer)]
    (is (= "<div></div>" (hs/sanitize s "<div><nope>Hello <span>world</span></nope></div>")))
    (hs/keep-child-nodes! s true)
    (is (true? (hs/keep-child-nodes? s)))
    (is (= "<div>Hello <span>world</span></div>"
           (hs/sanitize s "<div><nope>Hello <span>world</span></nope></div>")))))

(deftest test-10-on-removing-tag-cancels
  (with-open [s (hs/sanitizer)]
    (let [seen (atom [])]
      (hs/keep-tag-if! s (fn [node reason]
                           (swap! seen conj [(hs/node-name node) reason])
                           (= "keep-me" (hs/node-name node))))
      (is (= "<div><keep-me>a</keep-me></div>"
             (hs/sanitize s "<div><keep-me>a</keep-me><drop-me>b</drop-me></div>")))
      (is (some #{["keep-me" :not-allowed-tag]} @seen))
      (is (some #{["drop-me" :not-allowed-tag]} @seen)))))

(deftest test-11-on-filter-url-rewrites
  (with-open [s (hs/sanitizer)]
    (hs/rewrite-urls! s (fn [_ _ resolved]
                          (if (= "https://example.com/logo.png" resolved)
                            "https://cdn.example.net/logo.png"
                            resolved)))
    (is (= "<img src=\"https://cdn.example.net/logo.png\">"
           (hs/sanitize s "<img src=\"logo.png\">" "https://example.com")))))

(deftest test-12-handles-are-independent
  (with-open [a (hs/sanitizer)
              b (hs/sanitizer)]
    (hs/allow! a :tags "only-in-a")
    (is (hs/allowed? a :tags "only-in-a"))
    (is (not (hs/allowed? b :tags "only-in-a")))))

;; ---- extras: the remaining callback shapes -------------------------------

(deftest test-on-removing-attribute-sees-the-attribute
  (with-open [s (hs/sanitizer)]
    (let [seen (atom [])]
      (hs/keep-attribute-if! s (fn [elem attr _]
                                 (swap! seen conj [(hs/node-name elem)
                                                   (hs/attr-name attr)
                                                   (hs/attr-value attr)])
                                 false))
      (is (= "<div>x</div>" (hs/sanitize s "<div onclick=\"alert(1)\">x</div>")))
      (is (some #{["div" "onclick" "alert(1)"]} @seen)))))

(deftest test-on-removing-comment-cancels
  (with-open [s (hs/sanitizer)]
    (hs/keep-comment-if! s (constantly true))
    (is (= "<div>a<!-- keep -->b</div>" (hs/sanitize s "<div>a<!-- keep -->b</div>")))))

(deftest test-on-removing-style-is-four-arg
  (with-open [s (hs/sanitizer)]
    (let [seen (atom [])]
      (hs/keep-style-if! s (fn [_ nm v _]
                             (swap! seen conj [nm v])
                             (= "-custom-thing" nm)))
      (let [out (hs/sanitize s "<div style=\"-custom-thing: 3; color: red\">x</div>")]
        (is (.contains ^String out "-custom-thing"))
        (is (some #{["-custom-thing" "3"]} @seen))))))

(deftest test-post-process-node-visits
  (with-open [s (hs/sanitizer)]
    (let [kinds (atom [])]
      (hs/each-node! s (fn [node] (swap! kinds conj (hs/node-kind node))))
      (hs/sanitize s "<div><span>a</span><span>b</span></div>")
      (is (seq @kinds)))))

(deftest test-node-tree-navigation
  (with-open [s (hs/sanitizer)]
    (let [captured (atom nil)]
      (hs/each-document! s (fn [doc]
                             (reset! captured {:kind (hs/node-kind doc)
                                               :children (count (hs/children doc))})))
      (hs/sanitize s "<div>a</div><p>b</p>")
      (is (= :document (:kind @captured)))
      (is (>= (:children @captured) 2)))))

(deftest test-abi-version
  (with-open [s (hs/sanitizer)]
    (is (>= (hs/abi-version s) 1))))

(deftest test-sanitize-document-is-wired
  (with-open [s (hs/sanitizer)]
    (is (= "<div>doc</div>" (hs/sanitize-document s "<div>doc<script>x</script></div>")))))

(deftest test-attribute-set-value-rewrites
  (with-open [s (hs/sanitizer)]
    (hs/keep-attribute-if! s (fn [_ attr _]
                               (when (= "onclick" (hs/attr-name attr))
                                 (hs/set-attr-value! attr "safe")
                                 true)))
    (is (= "<div onclick=\"safe\">x</div>"
           (hs/sanitize s "<div onclick=\"alert(1)\">x</div>")))))

(deftest test-closed-sanitizer-rejects-use
  (let [s (hs/sanitizer)]
    (hs/close! s)
    (is (thrown? IllegalStateException (hs/sanitize s "<div>x</div>")))))

;; ---- extras specific to the Clojure layer --------------------------------

(deftest test-with-open-closes-even-when-the-body-throws
  ;; A leaked native handle would be invisible to every other check, so the
  ;; with-open contract gets one of its own.
  (let [captured (atom nil)]
    (is (thrown? IllegalStateException
                 (with-open [s (hs/sanitizer)]
                   (reset! captured s)
                   (throw (IllegalStateException. "boom")))))
    (is (thrown? IllegalStateException (hs/sanitize @captured "<div>x</div>")))))

(deftest test-sanitize-once
  (is (= "<div>Hello</div>" (hs/sanitize-once "<div onclick=\"alert(1)\">Hello</div>"))))

(deftest test-configure-applies-a-policy-map
  (with-open [s (hs/configure! (hs/sanitizer)
                               {:keep-child-nodes true
                                :allow {:tags ["my-widget"]}})]
    (is (hs/allowed? s :tags "my-widget"))
    (is (true? (hs/keep-child-nodes? s)))))

(deftest test-configure-installs-callbacks
  (with-open [s (hs/configure! (hs/sanitizer)
                               {:keep-tag-if (fn [n _] (= "keep-me" (hs/node-name n)))})]
    (is (= "<div><keep-me>a</keep-me></div>"
           (hs/sanitize s "<div><keep-me>a</keep-me><drop-me>b</drop-me></div>")))))

(deftest test-replace-allowed-starts-from-nothing
  (with-open [s (hs/sanitizer)]
    (hs/replace-allowed! s :tags "b" "i")
    (is (= 2 (hs/allow-count s :tags)))
    (is (= "<b>x</b>" (hs/sanitize s "<b>x</b><div>y</div>")))))

(deftest test-attribute-lookup-by-name
  (with-open [s (hs/sanitizer)]
    (let [found (atom nil)]
      (hs/keep-attribute-if! s (fn [elem _ _]
                                 (reset! found (some-> (hs/attribute elem "onclick")
                                                       hs/attr-value))
                                 false))
      (hs/sanitize s "<div onclick=\"alert(1)\">x</div>")
      (is (= "alert(1)" @found)))))

(deftest test-node->map-snapshots-a-borrowed-node
  ;; The point of node->map: the snapshot survives the callback, the Node does
  ;; not. Reading it AFTER sanitize returns is the whole check.
  (with-open [s (hs/sanitizer)]
    (let [snap (atom nil)]
      (hs/each-document! s (fn [doc] (reset! snap (hs/node->map doc))))
      (hs/sanitize s "<div>a</div><p>b</p>")
      (is (= :document (:kind @snap)))
      (is (>= (count (:children @snap)) 2))
      ;; Nothing here touches native memory — it is plain Clojure data.
      (is (every? map? (:children @snap))))))

(deftest test-walk-is-depth-first-and-inclusive
  (with-open [s (hs/sanitizer)]
    (let [n (atom 0)]
      (hs/each-document! s (fn [doc] (reset! n (count (hs/walk doc)))))
      (hs/sanitize s "<div><span>a</span></div>")
      ;; document + div + span + text, at least.
      (is (>= @n 3)))))

(deftest test-keep-predicate-tolerates-a-nil-return
  ;; Clojure truthiness, not unboxing: a fn falling off the end must mean
  ;; "do not keep", not NullPointerException.
  (with-open [s (hs/sanitizer)]
    (hs/keep-tag-if! s (fn [_ _] nil))
    (is (= "<div></div>" (hs/sanitize s "<div><nope>x</nope></div>")))))

(deftest test-rewrite-urls-treats-nil-as-no-rewrite
  (with-open [s (hs/sanitizer)]
    (hs/rewrite-urls! s (fn [_ _ _] nil))
    (is (= "<img src=\"https://example.com/logo.png\">"
           (hs/sanitize s "<img src=\"logo.png\">" "https://example.com")))))

(deftest test-unknown-reason-codes-degrade
  ;; The ABI's constants are append-only; a code this build has never seen must
  ;; become :unknown rather than blow up mid-sanitize.
  (is (= :unknown (hs/removal-reason 9999)))
  (is (= :unknown (hs/node-kind-of 9999))))

(deftest test-allow-list-rejects-an-unknown-selector
  (with-open [s (hs/sanitizer)]
    (is (thrown? IllegalArgumentException (hs/allow-list s :not-a-list)))))

;; ---- runner --------------------------------------------------------------

(defn -main
  "Entry point for `clojure.main -m`, so the suite runs with nothing but the
  Clojure jar. Exits non-zero if anything failed, which is what the aeb node
  keys off."
  [& _args]
  (let [{:keys [fail error]} (run-tests 'org.htmlsanitizer.conformance-test)]
    (shutdown-agents)
    (System/exit (if (pos? (+ fail error)) 1 0))))
